packages/gitbook/src/app/(space)/(core)/~gitbook/ogimage/[pageId]/route.tsx

```diff
@@ -1,53 +1,210 @@
+import { CustomizationHeaderPreset } from '@gitbook/api';
 import { redirect } from 'next/navigation';
 import { ImageResponse } from 'next/og';
-import { NextRequest } from 'next/server';
+import { ImageResponseOptions, NextRequest } from 'next/server';
+import colorContrast from 'postcss-color-contrast/js';
 import React from 'react';
 
+import { absoluteHref } from '@/lib/links';
+import { tcls } from '@/lib/tailwind';
 import { getContentTitle } from '@/lib/utils';
 
 import { PageIdParams, fetchPageData } from '../../../../fetch';
 
 export const runtime = 'edge';
 
+// TODO: Support all fonts available in GitBook
+// Right now this is impossible since next/font/google does not expose the cached font file
+// Another option would be to use the Satori prop `loadAdditionalAsset` [example](https://github.com/vercel/satori/blob/main/playground/pages/index.tsx),
+// but this prop isn't (yet) exposed through `ImageResponse`.
+const interRegular = fetch(
+    new URL('../../../../../../fonts/Inter/Inter-Regular.ttf', import.meta.url),
+).then((res) => res.arrayBuffer());
+const interBold = fetch(
+    new URL('../../../../../../fonts/Inter/Inter-Bold.ttf', import.meta.url),
+).then((res) => res.arrayBuffer());
+
+// By passing ImageResponse (Satori) a global object we make it cachable.
+// This also caches the font data, speeding up rendering.
+const imageOptions: ImageResponseOptions = {
+    width: 1200,
+    height: 630,
+    fonts: [
+        {
+            name: 'Inter',
+            data: await interRegular,
+            weight: 400,
+            style: 'normal',
+        },
+        {
+            name: 'Inter',
+            data: await interBold,
+            weight: 700,
+            style: 'normal',
+        },
+    ],
+};
+
 /**
  * Render the OpenGraph image for a space.
  */
 export async function GET(req: NextRequest, { params }: { params: PageIdParams }) {
     const { space, page, customization, parent } = await fetchPageData(params);
-    const url = new URL(space.urls.published ?? space.urls.app);
 
     if (customization.socialPreview.url) {
         // If user configured a custom social preview, we redirect to it.
         redirect(customization.socialPreview.url);
     }
 
+    const theme = customization.themes.default;
+    const light = theme === 'light';
+
+    const bold = customization.header.preset == CustomizationHeaderPreset.Bold;
+    const custom = customization.header.preset == CustomizationHeaderPreset.Custom;
+
+    // We have no access to CSS variables, so we'll have to hardcode some values
+    const baseColors = {
+        light: '#fff',
+        dark: '#111827',
+    };
+
+    const colors = {
```

Comment:
This logic is very messy, but it works. I’m intending to refactor this into more readable code before we merge.